### PR TITLE
Feature/vouchers

### PR DESCRIPTION
**Pull request by victazzz**

### CODE REVIEW

File: src/services/escola-lms/vouchers.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import { request } from 'umi';
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:1]: use requests from export https://escolalms.github.io/ts-models/modules/EscolaLms.Vouchers.Http.Requests.html

Comment (qunabu) [reply src/services/escola-lms/vouchers.ts:1]: Add this defs to all requests

**Review by qunabu — CHANGES_REQUESTED** [review]

File: src/components/ProductsSelect/index.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      {products.map((product) => (
+        <Select.Option key={product.id} value={product.id}>
+          {product.name}
+        </Select.Option>
+      ))}
+    </Select>
+  );
+};
+
+export default UserSelect;
```
Comment (qunabu) [inline src/components/ProductsSelect/index.tsx:90]:
```suggestion
export default ProductsSelect;
```

File: src/pages/Vouchers/form.tsx [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+const mapper = (item) => (typeof item === 'object' ? item.id : item);
+
+const VoucherForm = () => {
+  const intl = useIntl();
+  const params = useParams<{ voucherId?: string; tab?: string }>();
+  const { voucherId, tab = 'attributes' } = params;
+  const isNew = voucherId === 'new';
+
+  const [data, setData] = useState<Partial<EscolaLms.Vouchers.Models.Coupon>>();
```
[the remark this answers is not in the document]
Comment (qunabu) [reply src/pages/Vouchers/form.tsx:25]: up up

File: src/pages/Vouchers/index.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    title: <FormattedMessage id="active_to" defaultMessage="active_to" />,
+    dataIndex: 'active_to',
+    hideInSearch: true,
+    render: (_, record) => record.active_to && format(new Date(record.active_to), DAY_FORMAT),
+  },
+];
+
+const Vouchers: React.FC = () => {
+  const actionRef = useRef<ActionType>();
+  const [loading, setLoading] = useState(false);
```
Comment (qunabu) [inline src/pages/Vouchers/index.tsx:76]:
```suggestion
  const [loading, setLoading] = useState<boolean>(false);
```

File: src/services/escola-lms/vouchers.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+/**  GET /api/admin/vouchers */
+export async function vouchers(params?: API.PageParams, options?: Record<string, any>) {
+  return request<API.DefaultMetaResponse<EscolaLms.Vouchers.Http.Requests.ListCouponsRequest>>(
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:5]:
```suggestion
export async function vouchers(params?: API.PageParams & EscolaLms.Vouchers.Http.Requests.ListCouponsRequest, options?: RequestOptionsInit) {
  return request<API.DefaultMetaResponse<EscolaLms.Vouchers.Models.Coupon>>(
```

File: src/services/escola-lms/vouchers.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        per_page: params && params.pageSize,
+        page: params && params.current,
+      },
+      ...(options || {}),
+    },
+  );
+}
+
+/**  POST /api/admin/vouchers */
+export async function createVoucher(body?: Record<string, any>, options?: Record<string, any>) {
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:21]:
```suggestion
export async function createVoucher(body?: CreateCouponRequest, options?: RequestOptionsInit) {
```



in import parts 
```ts 
import type { RequestOptionsInit } from 'umi-request';
```

File: src/services/escola-lms/vouchers.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    data: body,
+    ...(options || {}),
+  });
+}
+
+/**  GET /api/admin/vouchers/:id */
+export async function getVoucher(id: number, options?: Record<string, any>) {
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:33]:
```suggestion
export async function getVoucher(id: number, options?: RequestOptionsInit) {
```

File: src/services/escola-lms/vouchers.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      ...(options || {}),
+    },
+  );
+}
+
+/**  PATCH /api/admin/vouchers */
+export async function updateVoucher(
+  id: number,
+  body?: Record<string, any>,
+  options?: Record<string, any>,
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:48]:
```suggestion
  body?: EscolaLms.Vouchers.Http.Requests. UpdateCouponRequest,
  options?: RequestOptionsInit,
```

File: src/services/escola-lms/vouchers.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        'Content-Type': 'application/json',
+      },
+      data: body,
+      ...(options || {}),
+    },
+  );
+}
+
+/**  DELETE /api/admin/vouchers/:id */
+export async function deleteVoucher(id: number, options?: Record<string, any>) {
```
Comment (qunabu) [inline src/services/escola-lms/vouchers.ts:64]:
```suggestion
export async function deleteVoucher(id: number, options?: RequestOptionsInit) {
```

File: src/services/escola-lms/webinars.ts [hunk 10]
```diff
@@ -21,7 +21,7 @@ export async function webinars(
 }
 
 /**  POST /api/webinars */
-export async function createWebinar(body?: Record<string, any>, options?: Record<string, any>) {
+export async function createWebinar(body?: Partial<API.Webinar>, options?: Record<string, any>) {
```
Comment (qunabu) [inline src/services/escola-lms/webinars.ts:24]:
```suggestion
export async function createWebinar(body?: Partial<API.Webinar>, options?: RequestOptionsInit) {
```

File: src/services/escola-lms/webinars.ts [hunk 11]
```diff
@@ -57,7 +57,7 @@ export async function updateWebinar(
   });
 }
 
-/**  GET /api/admin/webinars/:id */
+/**  DELETE /api/admin/webinars/:id */
 export async function deleteWebinar(id: number, options?: Record<string, any>) {
```
Comment (qunabu) [inline src/services/escola-lms/webinars.ts:61]:
```suggestion
export async function deleteWebinar(id: number, options?: RequestOptionsInit) {
```

**Review by qunabu — CHANGES_REQUESTED** [review]

File: src/pages/Vouchers/form.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (victazzz) [reply src/pages/Vouchers/form.tsx:25]: the id is missing from these types and I don't know if it is better to extend this type & {id: or use this type with models?

**Review by qunabu — APPROVED** [review]

**Pull request merged**